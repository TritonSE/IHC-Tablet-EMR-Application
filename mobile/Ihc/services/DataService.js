/* global */
// Returns either fakeLocalData or RealData depending on config file
// Config file:
// "FakeDataServices": "true" or "false"
import * as fakeLocalData from '../services/FakeLocalDataService';
import * as localDataService from '../services/LocalDataService';

import * as fakeServerData from '../services/FakeServerDataService';
import * as serverDataService from '../services/ServerDataService';

import config from '../config.json';

// Dont use the Fakes in tests, so that we can test the LocalDataService itself,
// but should stub out its methods for the other tests
let localData = (config.fakeDataServices === 'true') ? fakeLocalData : localDataService;
let serverData = (config.fakeDataServices === 'true') ? fakeServerData : serverDataService;
/*
let localData = (config.fakeDataServices === 'true' || global.__TEST__) ? fakeLocalData : localDataService;
let serverData = (config.fakeDataServices === 'true' || global.__TEST__) ? fakeServerData : serverDataService;
*/
export {localData, serverData};
