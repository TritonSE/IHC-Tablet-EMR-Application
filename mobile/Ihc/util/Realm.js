// Realm stores arrays as objects, so convert them before sending to server
export function convertPatientForServer(patient) {
  const copy = Object.assign({}, patient);
  copy.drugUpdates = Array.from(patient.drugUpdates);
  copy.soaps = Array.from(patient.soaps);
  copy.triages = Array.from(patient.triages);
  copy.statuses = Array.from(patient.statuses);
  return copy;
}

export function convertStatusForServer(statusObj) {
  const copy = Object.assign({}, statusObj);
  copy.medicationCheckmarks = Array.from(statusObj.medicationCheckmarks);
  return copy;
}
