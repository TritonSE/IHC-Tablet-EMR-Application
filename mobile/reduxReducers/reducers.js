import {combineReducers} from 'redux';
import * as containerActions from '../reduxActions/containerActions';

// individual reducers have a function name that will equal their property name
// in the state object
// i.e. state.loading will become equal to whatever this function returns
// The state parameter represents the preexisting state, and should be returned
// if the action does not match.
// If the action does match, then return the new value that is passed within the
// action.
function loading(state = false, action) {
  switch (action.type) {
    case containerActions.SET_LOADING:
      return action.loading;
    default:
      return state;
  }
}

function uploading(state = false, action) {
  switch (action.type) {
    case containerActions.IS_UPLOADING:
      return action.uploading;
    default:
      return state;
  }
}

function showRetryButton(state = false, action) {
  switch (action.type) {
    case containerActions.SET_LOADING:
      return action.showRetryButton;
    default:
      return state;
  }
}

function messages(state = {successMessage: null, errorMessage: null}, action) {
  switch (action.type) {
    case containerActions.SET_SUCCESS_MESSAGE:
      return {successMessage: action.message, errorMessage: null};
    case containerActions.SET_ERROR_MESSAGE:
      return {errorMessage: action.message, successMessage: null};
    case containerActions.CLEAR_MESSAGES:
      return {successMessage: null, errorMessage: null};
    default:
      return state;
  }
}

function currentPatientKey(state = null, action) {
  switch (action.type) {
    case containerActions.SET_CURRENT_PATIENT_KEY:
      return action.currentPatientKey;
    default:
      return state;
  }
}

const reducers = combineReducers({
  loading,
  uploading,
  showRetryButton,
  messages,
  currentPatientKey
});

export default reducers;
