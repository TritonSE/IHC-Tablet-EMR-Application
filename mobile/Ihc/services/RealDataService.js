// This file should hold all the fetch() calls to the Express server
// and the local database calls to the Realm DB
// TODO: Keep realm in sync with mongo
// All lastUpdated fields on mobile-side should be handled within this file.

import Patient from '../models/Patient';
import Status from '../models/Status';
import Soap from '../models/Soap';
import Triage from '../models/Triage';
import DrugUpdate from '../models/DrugUpdate';
import Settings from '../models/Settings';

import {createPatientHelper} from './functions/CreatePatient';
import {signinPatientHelper} from './functions/SigninPatient';
import {updateStatusHelper} from './functions/UpdateStatus';
import {updateSoapHelper} from './functions/UpdateSoap';
import {createDrugUpdateHelper} from './functions/CreateDrugUpdate';
import {getSoapHelper} from './functions/GetSoap';
import {getTriageHelper} from './functions/GetTriage';
import {getPatientHelper} from './functions/GetPatient';
import {getPatientSelectRowsHelper} from './functions/GetPatientSelectRows';
import {updateTriageHelper} from './functions/UpdateTriage';
import {getMedicationUpdatesHelper} from './functions/GetMedicationUpdates';
import {uploadUpdatesHelper} from './functions/UploadUpdates';
import {downloadUpdatesHelper} from './functions/DownloadUpdates';

import Realm from 'realm';
import {stringDate} from '../util/Date';
import config from '../config.json';

// Must set the fetchUrl to the server's IP Address and Port
const fetchUrl = config.fetchUrl;

const realm = new Realm({
  schema: [Patient, Status, Soap, Triage, DrugUpdate, Settings],
  deleteRealmIfMigrationNeeded: true, // TODO: delete when done with dev
});

export function createPatient(patient) {
  return createPatientHelper(patient, realm, fetchUrl);
}

// Check that patient exists, and if so then create a status object for them
export function signinPatient(patientForm) {
  return signinPatientHelper(patientForm, realm, fetchUrl);
}

// field: Any of fields from Status.schema
// value: should match the type that the Status.schema says
export function updateStatus(patientKey, strDate, field, value) {
  return updateStatusHelper(patientKey, strDate, field, value, realm, fetchUrl);
}

export function createDrugUpdate(update) {
  return createDrugUpdateHelper(update, realm, fetchUrl);
}

export function updateSoap(update) {
  return updateSoapHelper(update, realm, fetchUrl);
}

// Returns SOAP object if it exists, or undefined if not
export function getSoap(patientKey, strDate) {
  return getSoapHelper(patientKey, strDate, realm, fetchUrl);
}

export function updateTriage(update) {
  return updateTriageHelper(update, realm, fetchUrl);
}

// Returns Triage object if it exists, or undefined if not
export function getTriage(patientKey, strDate) {
  return getTriageHelper(patientKey, strDate, realm, fetchUrl);
}

export function getPatient(patientKey) {
  return getPatientHelper(patientKey, realm, fetchUrl);
}

/*
 * Return the statuses of the patients that are active and for this date
 */
export function getPatientSelectRows() {
  return getPatientSelectRowsHelper(realm, fetchUrl);
}

export function getMedicationUpdates(patientKey) {
  return getMedicationUpdatesHelper(patientKey, realm, fetchUrl);
}

export function uploadUpdates() {
  return uploadUpdatesHelper(realm, fetchUrl);
}

export function downloadUpdates() {
  return downloadUpdatesHelper(realm, fetchUrl);
}
