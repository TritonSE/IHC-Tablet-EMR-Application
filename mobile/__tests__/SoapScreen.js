import 'react-native';
import React from 'react';
import SoapScreen from '../screens/SoapScreen';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {localData} from '../services/DataService';
import sinon from 'sinon';
import Soap from '../models/Soap';

import { createStore } from 'redux';
import reducers from '../reduxReducers/reducers';
import { Provider } from 'react-redux';
const store = createStore(reducers);

import Realm, {mockObjects} from '../__mocks__/realm';
jest.mock('realm');

it('renders correctly', () => {
  // Calls downstreamSyncWithServer, so return mock Settings object
  mockObjects.mockImplementation(() => {
    return {lastSynced: 100};
  });

  sinon.stub(localData, 'getSoap').returns(Soap.getInstance());
  const json = renderer.create(
    <Provider store={store}>
      <SoapScreen todayDate='20180101' />
    </Provider>
  ).toJSON();
  expect(json).toMatchSnapshot();
});
