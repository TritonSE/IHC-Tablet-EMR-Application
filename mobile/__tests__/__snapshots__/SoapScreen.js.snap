// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#F5FCFF",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#F5FCFF",
        "flex": 1,
        "height": "100%",
        "justifyContent": "center",
        "width": "100%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "margin": 4,
        }
      }
    >
      Loading...
    </Text>
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="large"
    />
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#2196F3",
          "borderRadius": 2,
          "elevation": 4,
          "height": 40,
          "justifyContent": "center",
          "margin": 4,
          "opacity": 1,
          "padding": 8,
          "width": 120,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#fefefe",
              "fontWeight": "500",
              "textAlign": "center",
            },
            undefined,
          ]
        }
      >
        Cancel
      </Text>
    </View>
  </View>
</View>
`;
